language_service/src/rename.rs

```diff
@@ -0,0 +1,375 @@
+                                compilation: self.compilation,
+                                locations: vec![],
+                            };
+                            rename.visit_package(&self.compilation.unit.ast.package);
+                            rename.locations.push(def_span);
+                            self.locations = rename.locations;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn get_spans_for_local_rename(&mut self, node_id: ast::NodeId, ast_name: &ast::Ident) {
+        if let Some(curr) = self.current_callable {
+            if self.is_prepare {
+                self.prepare = Some((ast_name.span, ast_name.name.to_string()));
+            } else {
+                let mut rename = LocalRename {
+                    node_id,
+                    compilation: self.compilation,
+                    locations: vec![],
+                };
+                rename.visit_callable_decl(curr);
+                self.locations = rename.locations;
+            }
+        }
+    }
+}
+
+fn span_contains(span: Span, offset: u32) -> bool {
+    offset >= span.lo && offset <= span.hi
+}
+
+fn ast_item_id_to_hir_item_id(
+    item_id: ast::NodeId,
+    compilation: &Compilation,
+) -> Option<&hir::ItemId> {
+    if let Some(resolve::Res::Item(item_id)) = compilation.unit.ast.names.get(item_id) {
+        Some(item_id)
+    } else {
+        None
+    }
+}
+
+impl<'a> Visitor<'a> for Rename<'a> {
+    // Handles callable and UDT definitions
+    fn visit_item(&mut self, item: &'a ast::Item) {
+        if span_contains(item.span, self.offset) {
+            match &*item.kind {
+                ast::ItemKind::Callable(decl) => {
+                    if span_contains(decl.name.span, self.offset) {
+                        if let Some(item_id) =
+                            ast_item_id_to_hir_item_id(decl.name.id, self.compilation)
+                        {
+                            self.get_spans_for_item_rename(item_id, &decl.name);
+                        }
+                    } else if span_contains(decl.span, self.offset) {
+                        let context = self.current_callable;
+                        self.current_callable = Some(decl);
+                        walk_callable_decl(self, decl);
+                        self.current_callable = context;
+                    }
+                    // Note: the `item.span` can cover things like doc
+                    // comment, attributes, and visibility keywords, which aren't
+                    // things we want to have logic for, while the `decl.span` is
+                    // specific to the contents of the callable decl, which we do want
+                    // logic for. If the `if` or `else if` above is not met, then
+                    // the cursor is at one of these non-decl parts of the item,
+                    // and we want to do nothing.
```

Comment:
Oh, this comment is giving me some real deja-vu. Doesn't have to be this PR, but I really do think we should look at refactoring the visitor between definition, hover, and rename. (I did agree with your offline point about not all language features necessarily fitting in the same mold. But I think these three features are siblings, in a way, because they all deal with locating definitions and references. Find-All-References will be another one that's going to look very similar).

Reply:
Yah it is tough because all three are doing a similar step in that they need to find where the user cursor is located in the AST. The problem is that there is not a way to generalize over locations like that, and each of those features need to grab different information from those locations and transform them in different ways. Maybe this could be accomplished by using a set of callback functions, but that seems like a lot of callback functions. Maybe this could be accomplished with inheritance? Not sure right now.

Reply:
Discussed offline: design for new component discussed.